worlds/jakanddaxter/Rules.py

```diff
@@ -0,0 +1,222 @@
+            can_trade_orbsanity(state, player, world, required_orbs, required_previous_trade))
+
+
+def recalculate_reachable_orbs(state: CollectionState, player: int, world: JakAndDaxterWorld) -> None:
+
+    if not state.prog_items[player]["Reachable Orbs Fresh"]:
+
+        # Recalculate every level, every time the cache is stale, because you don't know
+        # when a specific bundle of orbs in one level may unlock access to another.
+        for level in level_table:
+            state.prog_items[player][f"{level} Reachable Orbs".strip()] = (
+                count_reachable_orbs_level(state, world, level))
+
+        # Also recalculate the global count, still used even when Orbsanity is Off.
+        state.prog_items[player]["Reachable Orbs"] = count_reachable_orbs_global(state, world)
+        state.prog_items[player]["Reachable Orbs Fresh"] = True
+
+
+def count_reachable_orbs_global(state: CollectionState,
+                                world: JakAndDaxterWorld) -> int:
+
+    accessible_orbs = 0
+    for level_regions in world.level_to_regions.values():
+        for region in level_regions:
+            # Rely on short-circuiting to skip region.can_reach whenever possible.
+            if region.orb_count > 0 and region.can_reach(state):
+                accessible_orbs += region.orb_count
+    return accessible_orbs
+
+
+def count_reachable_orbs_level(state: CollectionState,
+                               world: JakAndDaxterWorld,
+                               level_name: str = "") -> int:
+
+    accessible_orbs = 0
+    for region in world.level_to_regions[level_name]:
+        # Rely on short-circuiting to skip region.can_reach whenever possible.
+        if region.orb_count > 0 and region.can_reach(state):
+            accessible_orbs += region.orb_count
+    return accessible_orbs
+
+
+def can_reach_orbs_global(state: CollectionState,
+                          player: int,
+                          world: JakAndDaxterWorld,
+                          bundle: int) -> bool:
+
+    recalculate_reachable_orbs(state, player, world)
+    return state.has("Reachable Orbs", player, world.orb_bundle_size * (bundle + 1))
+
+
+def can_reach_orbs_level(state: CollectionState,
+                         player: int,
+                         world: JakAndDaxterWorld,
+                         level_name: str,
+                         bundle: int) -> bool:
+
+    recalculate_reachable_orbs(state, player, world)
+    return state.has(f"{level_name} Reachable Orbs", player, world.orb_bundle_size * (bundle + 1))
+
+
+def can_trade_vanilla(state: CollectionState,
+                      player: int,
+                      world: JakAndDaxterWorld,
+                      required_orbs: int,
+                      required_previous_trade: typing.Optional[int] = None) -> bool:
+
+    # With Orbsanity Off, Reachable Orbs are in fact Tradeable Orbs.
+    recalculate_reachable_orbs(state, player, world)
```

Comment:
For what it's worth, in Python, function calls do actually cost a somewhat significant amount of time. Obviously I'm not saying you should inline this whole function everywhere, but I think it'd be nice if the "stale check" was actually here and not inside the function

It's ok if you don't want to though

Reply:
Yep, function calls have overhead. With such a simple condition check, and all the code changes limited to one file, and if it can avoid calling a function unnecessarily, that's worth doing.